---
title: "SNPs of Interest - Part II"
author: "FiG-T"
data: "`r Sys.Date()`"
output: 
  html_document: 
    keep_md: yes
    toc: yes
toc: TRUE
editor_options:
  markdown:
    wrap: 80
  chunk_output_type: inline
---

# Introduction

```{r libraries}

library(ggplot2)
library(stringr)
library(dplyr)
library(tidyr)
#library(tidytree)
library(ape)

```

### Palettes

Add in source information -------------------------------------------

```{r}

FiGT_continent_palette <- c(
  Africa = "deepskyblue4",
  Asia = "firebrick4", 
  Europe ="orchid3", 
  `North America` ="chartreuse4",
  `South America` = "chartreuse3",
  Oceania = "turquoise"
  )

```

## Required Data

```{r required_data}

# load in paleo info
paleo_wcvp_outbreaks <- feather::read_feather(
  path = "~/Documents/data/lhf_d/feather/lhf_meta_paleo_wcvp_outbreaks_05_2024.feather"
)

# import midpoint rooted tree (see above)
FastTree_full_rooted <- ape::read.tree(
  file = "~/Documents/data/lhf_d/trees/mafft_aln_filtered_FastTree_02_2024_rooted.tree"
)

lhf_gt_pc_st2_meta <- feather::read_feather(
  path = "~/Documents/data/lhf_d/feather/lhf_gt_meta_pc_ds_subset_full_08_2024.feather"
)

lhf_gt_meta_ds_subset <- feather::read_feather(
  path = "~/Documents/data/lhf_d/feather/lhf_gt_meta_pc_ds_subset_08_2024.feather"  
)

lhf_names <- readr::read_delim(
  file = "~/Documents/data/lhf_d/vcf/lhf_filtered_maf0.005_sampleNames.txt", 
  delim = "/t", 
  col_names = "samples"
)

lhf_gt_meta_ds_subset_core_hits <- feather::read_feather(
  path = "~/Documents/data/lhf_d/feather/lhf_gt_meta_ds_subset_core_hits_08_2024.feather" # date changed
)
```

# SNP linkage

Mitochondrial DNA has often been considered as a single locus (due to the lack
of recombination). A possible criticism of this work is therefore that all
detected sites are highly correlating with each other and I do not have multiple
independent signals.

LD is hard to calculate so a simple R[^1] has been used instead.

[^1]: 2

```{r calculate_SNP_cor}

GT_cor <- Hmisc::rcorr(
  as.matrix(
    #lhf_gt_pc_st2_meta[,c(43:ncol(lhf_gt_pc_st2_meta))]
    lhf_gt_meta_ds_subset_core_hits[,c(24:ncol(lhf_gt_meta_ds_subset_core_hits))]
  )
)

#GT_cor$r[lower.tri(GT_cor$r)] <- NA

GT_cor <- reshape2::melt(
  GT_cor$r,
  na.rm = TRUE
)

GT_cor$value <- round(
  x = GT_cor$value, 
  digits = 3
)

```

```{r change_labels}
# to match the rCRS positions

GT_cor_pos <- GT_cor %>%
  filter( str_detect(Var1, "pos_.*")) %>%
  select(Var1) %>%
  tidyr::separate(
    Var1, 
    c("text","pos"), 
    sep = "_"
  ) %>%
  distinct

GT_cor_pos$pos <- as.numeric(GT_cor_pos$pos)

GT_cor_pos <- left_join(
  GT_cor_pos, 
  rCRS, 
  by = c("pos" = "aln_pos")
)
 
GT_cor_pos$rCRS_pos <- stringr::str_c(
  "pos_", 
  GT_cor_pos$rCRS_pos
)
GT_cor_pos$pos <- stringr::str_c(
  "pos_", 
  GT_cor_pos$pos
)
 
for(i in seq_len(nrow(GT_cor_pos))) {
  GT_cor <- GT_cor %>%
    mutate(
      Var1 = str_replace_all(
        Var1, 
        pattern = GT_cor_pos$pos[i], 
        replacement = GT_cor_pos$rCRS_pos[i]
      )
    )
  
  GT_cor <- GT_cor %>%
    mutate(
      Var2 = str_replace_all(
        Var2, 
        pattern = GT_cor_pos$pos[i], 
        replacement = GT_cor_pos$rCRS_pos[i]
      )
    )
}

GT_cor_env <- GT_cor %>%
  filter( ! str_detect(Var1, "pos_.*")) %>%
  select(Var1) %>%
  distinct() 

GT_cor_env <- c(GT_cor_env$Var1)

GT_cor$Var1 <- factor(
  x = GT_cor$Var1, 
  levels = c(
    GT_cor_env,
    c(GT_cor_pos$rCRS_pos)
  )
)

GT_cor$Var2 <- factor(
  x = GT_cor$Var2, 
  levels = c(
    GT_cor_env,
    c(GT_cor_pos$rCRS_pos)
  )
)


#levels(GT_cor$Var2)
```

```{r plot_correlations}
ggplot2::ggplot(
  data = GT_cor,
  ggplot2::aes(
    x = Var2,
    y = Var1,
    fill = value
  )
) +
  geom_tile(
    colour = "white"
  ) +
#  geom_text(
#    mapping = aes(
#      label = value
#    ), 
#    #size = 1.25 # if saving to pdf
#    size = 2.5
#  ) +
  xlab("")+
  ylab("")+
  scale_fill_gradient2(
    low = "deepskyblue4",
    high = "darkorchid3",
    mid = "snow3",
    midpoint = 0,
    #limit = c(0.5,1),
    #space = "Lab",
    #name="Pearson\nCorrelation"
  ) +
  labs(fill = "R-squared Correlation Coefficient") +
  theme(
    axis.text.x = element_text(
      angle = 90,
      vjust = 0.25,
      hjust = 0.5,
      #size = 12, 
      size = 6
    ), 
    axis.text.y = element_text(
      angle = 0,
      vjust = 0.5,
      hjust = 0.5, 
      #size = 12, 
      size = 6
    ), 
    legend.position = "bottom"
    # axis.title = element_blank(), 
    # legend.position = "none"
  ) #+
  #transparent_theme
```

Note that these SNPs are phased to the rCRS.

```{r save_SNP_correlation_plot}

ggplot2::ggsave(
  filename = "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/figures/supplementary/lhf_core_GLM_hits_correlation_heatmap_08_2024_08_2024.png",
  plot = last_plot(),
  bg = "transparent", 
  width = 12, 
  height = 7
)

```

## Results

From initial inspection:

-   All the environmental variable correlate reasonably well with each other.

-   NPP and LAI correlate with precipitation variables (NPP more so than LAI)

-   SNPs within the mtDNA membrane Attachment Site (ATT) correlate strongly with
    each other (and are thus likely linked)

-   Blocks of SNPs correlate - these are not necessarily adjacent. *Are these
    indicative of haplogroups/haplogroup markers.*

-   Some SNPs do correlate with select environmental variables (see box below)

> SNPs that correlate:
>
> 8701: link with T_wetQ_paleo
>
> 10238: correlates (+vely) with precipitation variables
>
> 8251: correlates (-vely) with precipitation variables
>
> 3010: -ve correlation with temp variables

# Repeat mutations

```{r sorting_tree_names}
lhf_names <- lhf_names %>%
  separate_wider_delim(
    cols = samples, 
    names = c("acc", "country"),
    too_many = "merge",
    delim = "_", 
    cols_remove = FALSE
  )

names(lhf_names) <- c("acc", "country_og", "samples_og")

glm_indv_acc <- lhf_gt_meta_ds_subset_core_hits %>% 
  #filter(!is.na(pos_93)) %>%
  select(acc)

glm_og_links <- left_join(
  x = glm_indv_acc, 
  y = lhf_names
)

glm_indv <- c(glm_og_links$samples_og)
length(glm_indv)
```

```{r pruning_tree}

pruned_tree <- drop.tip(
  FastTree_full_rooted, 
  setdiff(
    FastTree_full_rooted$tip.label, 
    glm_indv
  )
)

length(FastTree_full_rooted$tip.label)
length(pruned_tree$tip.label)

# create a new tibble for the tree metadata 
tree_meta <- dplyr::as_tibble(pruned_tree$tip.label) %>% distinct()
names(tree_meta) <- "samples_og"

tree_meta <- dplyr::left_join(
  tree_meta, 
  glm_og_links,  # this is queried above
  by = join_by("samples_og"=="samples_og")
)

# combine populations (countries) with the associated metadata
tree_meta <- dplyr::left_join(
  tree_meta, 
  lhf_gt_meta_ds_subset,  # this is queried above
  by = join_by("acc"=="acc")
)
```

```{r plotting_trees}

pdf(
  file = "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/figures/supplementary/lhf_hits_trees_08_2024.pdf" , 
  onefile = TRUE, 
  width = 9, 
  height = 9
)

 allele_palette <- c(
    "black", "maroon2", "turquoise2", "forestgreen"
  )

for (i in seq_along(GT_cor_pos$pos) ) { # see above for GT_cor_pos
  C
  j <- GT_cor_pos$pos[i]
  k <- GT_cor_pos$rCRS_pos[i]
  
  ape::plot.phylo(
    x = pruned_tree, 
    type = "fan",
    show.tip.label = FALSE,
    #use.edge.length = FALSE, 
    cex = 0.7, 
    #srt = 20, 
    #label.offset = 0.05, 
    edge.color = allele_palette[as.numeric(
      as.factor(
        tree_meta[[j]]
      )
    )], 
    main = k
  )
  
}

dev.off()

GT_cor_pos$pos

```

```{r continental_tree}
unique(as.factor(tree_meta[["continent"]]))

pruned_tree_countries <- pruned_tree 
pruned_tree_countries$tip.label <- tree_meta$country

length(pruned_tree_countries$tip.label)
length(tree_meta$country)

pdf(
  file = "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/figures/supplementary/lhf_indv_continental_tree_08_2024.pdf" , 
  onefile = TRUE, 
  width = 9, 
  height = 9
)

ape::plot.phylo(
  x = pruned_tree_countries, 
  type = "fan",
  show.tip.label = FALSE,
  #use.edge.length = FALSE, 
  cex = 0.2, 
  srt = 20, 
  label.offset = 0.01, 
  edge.color = FiGT_continent_palette[as.numeric(
    as.factor(
      tree_meta[["continent"]]
    ) 
  )], 
  main = "Continental Tree"
)
  
dev.off()

```

# Allele Frequencies

As well as looking at correlation/ linkage blocs between SNPs, I want to
understand *where* each SNP is common - this will help inform us about what is
the likely selective pressure and aid understanding of why a specific SNP may be
relevant.

## Filtering to only include hits

```{r filter_st2_meta}
st2_hits <- sort(
  unique(
    lhf_st2_GLMM_min$snp_pos #N.B use the non rCRS names
  )
)
st2_hits <- str_c(
  "pos_", st2_hits
)
st2_hits_rCRS <- sort(
  unique(
    lhf_st2_GLMM_min$rCRS_pos #N.B use the non rCRS names
  )
)
st2_hits_rCRS <- str_c(
  "pos_", st2_hits_rCRS
)


lhf_gt_pc_st2_meta_hits <- lhf_gt_pc_st2_meta %>%
  select(
    acc:lai_paleo, 
    st2_hits
  )
```

## Calculating Allele Frequency by Country

```{r calculate_AF}

# convert to long format 
lhf_AF_long <- pivot_longer(
  data = lhf_gt_pc_st2_meta_hits, 
  cols = matches("pos_.*"), 
  names_to = "position", 
  values_to = "genotype"
)

# group by country and SNP
lhf_AF_long <- lhf_AF_long %>%
  group_by(
    iso3, position 
  ) %>%
  #filter(genotype %in% c(0, 1) ) %>%  # only include first 2 alleles
  reframe(
    continent = continent,
    maf = mean(genotype)
  ) %>%
  distinct()
```

Use the labels defined above to replace SNP positions with rCRS aligned values

```{r rename_pos_af}

for(i in seq_along(st2_hits)) {
  
  lhf_AF_long <- lhf_AF_long %>%
    mutate(
      position = str_replace_all(
        position, 
        pattern = st2_hits[i], 
        replacement = st2_hits_rCRS[i]
      )
    )
  
}

lhf_AF_long$position <- factor(
  x = lhf_AF_long$position, 
  levels = c(st2_hits_rCRS)
  
)

#levels(lhf_AF_long$position)
```

```{r quick_plot_af, warning=FALSE}

for (i in levels(lhf_AF_long$position)) {
  
  data <- lhf_AF_long %>%
    filter(position == i)
  
  print(
    ggplot(
      data = data, 
        mapping = aes(
          x = reorder(x = iso3, +maf), 
          y = maf, 
          fill = continent
       )
    ) + 
    geom_col(
    ) +
    scale_fill_manual(
      values = FiGT_continent_palette
    ) +
    ylab(
      paste0(
        "Allele Frequency for ", i
      )
    ) +
    theme(
      axis.text.x = element_text(
        angle = 90, 
        size = 10
      )
    ) +
    xlab(
      "Country (ISO3 code)"
    ) +
    ylim(
      c(0,1)
    ) #+
    #theme_light()
  )
}

```

It would be interesting to see allele frequencies vs environmental metrics:

```{r format_metadata}

lhf_AF_long <- left_join(
  x = lhf_AF_long,
  y = lhf_meta2 %>%
    select(long:precip_seasonality) %>%
    select(-continent)
    distinct(),
  by = c("iso3")
)

lhf_AF_long <- left_join(
  x = lhf_AF_long,
  y = paleo_wcvp_outbreaks %>%
    distinct(),
  by = c("iso3")
)

```

```{r af_env_plotter, warning=FALSE, message=FALSE}

af_env_plotter <- function(
        input, 
        positions, 
        scores_input
) {
  
  # load in data 
  data <- input
  
  # load in score data
  scores <- scores_input
  
  scores$labels <- str_c(
    "pos_", 
    scores$rCRS_pos
  )
  
  for (i in positions) {
    
    # filter for a specific position
    data_filtered <- data %>%
      filter(position == i)
    
    scores_filtered <- scores %>%
      filter(labels == i )
    
    env_var <- unique(scores_filtered$covariate)
    
    #message(env_var)
    
    plot <- ggplot(
      data = data_filtered, 
        mapping = aes(
          y = maf, 
          x = .data[[env_var]] , 
          fill = continent, 
          colour = continent
       )
      ) + 
      #geom_smooth(
        #inherit.aes = FALSE,
      #  mapping = aes(
      #    x = maf, 
      #    y = .data[[env_var]]
      #  ), 
      #  method = 'lm', 
      #  #colour = "black", 
      #  se = FALSE,
      #  alpha = 0.75
      #) +
      geom_smooth(
        inherit.aes = FALSE,
        mapping = aes(
          y = maf, 
          x = .data[[env_var]]
        ), 
        method = 'lm', 
        colour = "black", 
        #se = FALSE,
        alpha = 0.75
      ) +
      geom_point(
        size = 3
      ) +
    # geom_text(
    #    mapping = aes(
    #      label = iso3
    #    ), 
    #    nudge_x = 1, 
    #    size = 2
    #  ) +
      scale_fill_manual(
        values = FiGT_continent_palette
      ) +
      scale_colour_manual(
        values = FiGT_continent_palette
      ) +
      ylab(
        paste0(
          "Allele Frequency for ", i
        )
      ) +
      theme(
        axis.text.x = element_text(
          angle = 90, 
          size = 3.5
        )
      ) +
      xlab(
        env_var
      ) +
      theme_light()
  
    # print plot
    
    print(plot)
  
  } # close loop
  
}

```

```{r af_env_plots, warning=FALSE}

pdf(
  file =  "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/figures/stage2_plots/lhf_GLMM_EnvvsAF_scatters_08_2024.pdf", 
  width = 8, 
  height = 6, 
  onefile = TRUE
)


af_env_plotter(
  input = lhf_AF_long, 
  positions = st2_hits_rCRS, 
  scores_input = lhf_st2_GLMM_min
)

dev.off()
```

```{r check_overlap_snps}
# combine core and stage 2 results 
combined_st1_st2_GLMs <- full_join(
  x = core_GLM_min_null, 
  y = lhf_st2_GLMM_min %>%
    select(
      rCRS_pos, covariate
    ) %>%
    distinct(), 
  by = "rCRS_pos"
)

core_st2_hits <- sort(c(unique(combined_st1_st2_GLMs$rCRS_pos)))

st2_hits <- c(unique(lhf_st2_GLMM_min$rCRS_pos))

# select the positions where there is no significant stage 2 hit
combined_st1_st2_GLMs <- combined_st1_st2_GLMs %>%
  filter(is.na(null_chisq)) %>%
  select(rCRS_pos) %>%
  distinct()

sort(combined_st1_st2_GLMs$rCRS_pos)

openxlsx::write.xlsx(
  x = combined_st1_st2_GLMs, 
  file = "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/combined_st1_st2_hits_08_2024.xlsx"
)
```

This shows that there are 14 SNPs (199, 239, 3918, 4580, 5046, 5460, 6455, 6734,
8701, 9947, 9950, 11944, 12414 & 14212) which are not detected as being
significant in the second stage of modelling. This could be as the Bonferroni
corrected threshold is higher (as there are many more tests) or that the
interact between latitude and precipitation is specifically important in the
core hits.

Three sites (rCRS 980, 9667 & 9932) are significant in the second stage of hits
but were not detected in the core models.

## Looking Functional Effects

Once the GLMMs have been calculated, the possible functional consequences of the
SNPs can be assessed. One approach to this is to pair up the SNPs with ENSEMBL's
Predicted Effects (ensembl.org). These predicted effects will be generated for
hits from both the core models and stage 2

## Creating a VCF for the GLMM hits

```{r list_snp_pos_hits}

snp_pos_hits <- sort(
  c(
    unique(
      combined_st1_st2_GLMs$snp_pos
      ), 
    1024,9734, 9999
    )
  )

snp_pos_hits <- str_c(
  1, 
  snp_pos_hits, 
  sep = " "
)

write.table(
  x = snp_pos_hits, 
  sep = "\t", 
  row.names = FALSE, 
  file = "~/Documents/data/lhf_d/vcf/GLMM_hits_snp_pos_08_2024.txt"
)
# find and replace " with blank in the output file 
# check that the CHROM column says 1 (for now)
```

> Check that the CHROM is corrected to specify that this is the mitochondrial
> genome

VCFtools can once again be used to filter the original VCF to only contain sites
that were detected as significant in the GLMMs.

```{bash}

# select only the required postions from the VCF
vcftools --vcf snp_sites_filtered_vcf_maf0.005_noRef_092023.recode.vcf --positions GLMM_hits_snp_pos_08_2024.txt --recode --out snp_pos_core_st2_GLMM_hits_08_2024 

# manually change the positions to match the reference genome (there must be a better way of doing this)

# compress this file 
gzip --keep rCRS_pos_core_st2_GLMM_hits_08_2024.vcf
```

These two files can then be used in the SIFT database (protein coding regions
only) and for the ENSEMBL VEP (all regions). For more information on how to
format these files please see MITOMAP_mutations.Rmd.

The consensus sequence can also be used to extract sequences for a specific
region.

```{r generate_specific_sequence}

# load in the consensus sequence (from the alignment)
consenus_sequence <- read.delim(
  file = "~/Documents/data/lhf_d/fasta/lhf_multiple_alignment_consensus.txt", 
  header = FALSE
)

# convert this into a long vector
consenus_sequence <- str_split(
  consenus_sequence, 
  boundary("character")
)
```

```{r specific_sequences}

# copy to a new sequence section
rRNA_16S <- consenus_sequence

# add in your alternate (if required)
rRNA_16S[[1]][3060] <- "a"

# filter to include the alignment positions
rRNA_16S <- rRNA_16S[[1]][c(1715:3281)]

# write in a flat string to an output file
write(
  x = str_flatten(rRNA_16S), 
  file = "~/Documents/data/lhf_d/fasta/rRNA_16S_consensus_mut3010.txt", 
  sep = ""
)

```

## Combining Hits and Predicted effects

```{r vep}

vep <- read.csv(
  file = "~/Documents/data/lhf_d/ensembl_VEP_formatted_specific.csv"
)

lhf_GLMM_vep <- left_join(
  x = combined_st1_st2_GLMs, 
  y = vep, 
  by = join_by(
    "rCRS_pos" == "pos" 
  )
)

write.csv(
  x = lhf_GLMM_vep, 
  file = "/Users/finleythomas/Library/CloudStorage/OneDrive-UniversityCollegeLondon/Results/lhf/lhf_GLMM_hits_loci_VEP_08_2024.csv"
)

```

```{r}
lhf_GLMM_vep_summary <- lhf_GLMM_vep %>%
  group_by(Consequence, classification) %>%
  reframe(
    N = n()
  )

ggplot2::ggplot(
  data = lhf_GLMM_vep_summary %>%
    filter(!is.na(Consequence)), 
  mapping = aes(
    x = Consequence, 
    y = N, 
    fill = Consequence
  )
) +
  geom_col(
    position = position_dodge2(0.2)
  ) +
  facet_grid(
    cols = vars(classification)
  )
```

```{r isolate_seq}
interest_fasta <- seqinr::read.fasta(
  file = "~/Documents/data/lhf_d/fasta/aln_mafft_incRef_filtered_3_09_2023.fasta"
)

interest_fasta <- bind_rows(interest_fasta)

str(tRNA_TT)

tRNA_TT <- interest_fasta[c(15955:16020),]

tRNA_TT <- t(tRNA_TT)

# convert to a tibble 
tRNA_TT <- tidyr::as_tibble(
  tRNA_TT,
  rownames = "samples" # retain the rownames (ids) and place into a column. 
)

tRNA_TT <- tRNA_TT %>%
  unite(
    sequence,
    V1:names(tRNA_TT)[ncol(tRNA_TT)], 
    sep = ""
  )

tRNA_TT <- tRNA_TT %>%
  select(sequence) %>%
  distinct()

```
